jubilant/secrettypes.py

```diff
@@ -0,0 +1,106 @@
+"""Dataclasses that contain parsed output from juju secrets."""
+
+from __future__ import annotations
+
+import dataclasses
+import datetime
+from typing import Dict, Generic, Literal, TypedDict, TypeVar
+
+from typing_extensions import NotRequired
+
+
+class SecretURI(str):
+    """A string subclass that represents a secret URI ("secret:...")."""
+
+    @property
+    def unique_identifier(self) -> str:
+        """Unique identifier of this secret URI.
+
+        This is the secret's globally-unique identifier (currently a 20-character Xid,
+        for example "9m4e2mr0ui3e8a215n4g").
+        """
+        if '/' in self:
+            # Handle 'secret://MODEL-UUID/UNIQUE-IDENTIFIER'
+            return self.rsplit('/', maxsplit=1)[-1]
+        elif self.startswith('secret:'):
+            # Handle common case of 'secret:UNIQUE-IDENTIFIER'
+            return self[len('secret:') :]
+        else:
+            return str(self)
+
+
+@dataclasses.dataclass
+class SecretAccess:
+    """Represents access to a secret."""
+
+    target: str
+    scope: str
+    role: str
```

Comment:
I think there are only 2 possible values for role, so it would be better as an enum (or even Literal). However, Ben might be less keen on that, so feel free to leave as-is and we can confirm when he's back.